### PR TITLE
Focus widget 2021

### LINKED ISSUE #924: Restore Focus Widget in the GUI
**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
Restore the Focus widget using information from BrainFlow metric. Front end can remain the same for now. Important thing is to have a Focus widget with modest accuracy.

Restore Focus Widget output to the Networking widget for all networking modes.

**Describe alternatives you've considered**
This was removed as per a team decision related to the accuracy of the old Focus widget.

**Additional context**
The backend work for this issue has been completed! BrainFlow now has a metric that uses a simple classifier. 🛠✅

Need to update Docs also once this is coded and tested in the GUI. Specifically, the GUI Widget Guide.

### PR DESCRIPTION
**Pull request by retiutut**
Fixes  #924

### CODE REVIEW

File: OpenBCI_GUI/W_Focus.pde [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+            int window_size = currentBoard.getSampleRate() * xLimit.getValue();
+            List<double[]> currentData = currentBoard.getData(window_size);
+            double[][] data = new double[currentData.size()][];
+            if (currentData.size() == 0) {
+                println("OOPS!!!!");
+                return -1d;
+            }
+            data = currentData.toArray(data);
+            println(data.length);
+            Pair<double[], double[]> bands = DataFilter.get_avg_band_powers (data, currentBoard.getEXGChannels(), currentBoard.getSampleRate(), false);
```
Comment (retiutut) [inline OpenBCI_GUI/W_Focus.pde:273]: @Andrey1994 

```
OpenBCI_GUI: Init session using Synthetic data source
Channel count set to 8
Need to call beginShape() first
InitSettings: Saving Default Settings to file!
SessionSettings: 3 active widgets saved!
[SUCCESS]: Session started!
[DEFAULT]: openBCI_GUI: startButton was pressed. Starting data transfer, wait a few seconds.
Brainflow start streaming
[DEFAULT]: Data stream started.
OpenBCI_GUI: openNewLogFile: opened ODF output file: /Users/Richard/Documents/OpenBCI_GUI/Recordings/OpenBCISession_2021-03-21_22-38-52/OpenBCI-RAW-2021-03-21_22-38-56.txt
brainflow.BrainFlowError: Failed to get_avg_band_powers:INVALID_ARGUMENTS_ERROR
	at brainflow.DataFilter.get_avg_band_powers(DataFilter.java:494)
	at OpenBCI_GUI$W_Focus.updateFocusState(OpenBCI_GUI.java:18762)
	at OpenBCI_GUI$W_Focus.update(OpenBCI_GUI.java:18634)
	at OpenBCI_GUI$WidgetManager.update(OpenBCI_GUI.java:26229)
	at OpenBCI_GUI.systemUpdate(OpenBCI_GUI.java:941)
	at OpenBCI_GUI.draw(OpenBCI_GUI.java:564)
	at processing.core.PApplet.handleDraw(PApplet.java:2482)
	at processing.opengl.PSurfaceJOGL$DrawListener.display(PSurfaceJOGL.java:866)
	at jogamp.opengl.GLDrawableHelper.displayImpl(GLDrawableHelper.java:692)
	at jogamp.opengl.GLDrawableHelper.display(GLDrawableHelper.java:674)
	at jogamp.opengl.GLAutoDrawableBase$2.run(GLAutoDrawableBase.java:443)
	at jogamp.opengl.GLDrawableHelper.invokeGLImpl(GLDrawableHelper.java:1293)
	at jogamp.opengl.GLDrawableHelper.invokeGL(GLDrawableHelper.java:1147)
	at com.jogamp.newt.opengl.GLWindow.display(GLWindow.java:759)
	at com.jogamp.opengl.util.AWTAnimatorImpl.display(AWTAnimatorImpl.java:81)
	at com.jogamp.opengl.util.AnimatorBase.display(AnimatorBase.java:452)
	at com.jogamp.opengl.util.FPSAnimator$MainTask.run(FPSAnimator.java:178)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
ERROR UPDATING FOCUS STATE!

```

**Comment by retiutut** [comment]
- [ ] Add channel select feature

File: OpenBCI_GUI/W_Focus.pde [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+        int ty = y + padding_5 + extraPadding;
+        int tw = int(upperLeftContainerW) - padding_5*2;
+        //tableHeight = tw;
+        dataGrid.setDim(tx, ty, tw);
+        dataGrid.setTableHeight(int(upperLeftContainerH - padding_5*2));
+        dataGrid.dynamicallySetTextVerticalPadding(0, 0);
+        dataGrid.setHorizontalCenterTextInCells(true);
+    }
+
+    private void update_crystalball_dims() {
```
Comment (Andrey1994) [inline OpenBCI_GUI/W_Focus.pde:189]: fix name to match code style

File: OpenBCI_GUI/W_Focus.pde [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+        float upperLeftContainerW = w/2;
+        float upperLeftContainerH = h/2;
+        float min = min(upperLeftContainerW, upperLeftContainerH);
+        xc = x + w/4;
+        yc = y + h/4;
+        wc = min * (3f/5);
+        hc = wc;
+    }
+
+    private void update_graph_dims() {
```
Comment (Andrey1994) [inline OpenBCI_GUI/W_Focus.pde:200]: and here

File: OpenBCI_GUI/W_Focus.pde [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+    private double updateFocusState() {
+        // todo move concentration.prepare() and variable initialization outside from this method, it should be called only once!
+        try {
+            int window_size = currentBoard.getSampleRate() * xLimit.getValue();
+            // getData in GUI returns data in shape ndatapoints x nchannels, in BrainFlow its transposed
+            List<double[]> currentData = currentBoard.getData(window_size);
+            if (currentData.size() != window_size || focusChanSelect.activeChan.size() <= 0) {
+                return -1.0;
+            }
+            int[] exgChannels = currentBoard.getEXGChannels();
```
Comment (Andrey1994) [inline OpenBCI_GUI/W_Focus.pde:217]: lets move exgChannels init from this method to smth like constructor, and probably it should be getEEGChannels(if board object has such method)

File: OpenBCI_GUI/W_Focus.pde [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+            int channelCount = currentBoard.getNumEXGChannels();
+            int[] channelsInDataArray = new int[channelCount];
+            /*
+            for (int j = 0; j < bpChanSelect.activeChan.size(); j++) {
+                int chan = bpChanSelect.activeChan.get(j);
+                */
+
+            int activeChanCounter = 0;
+            for (int i = 0; (i < channelCount) && (activeChanCounter < focusChanSelect.activeChan.size()); i++) // use this line to use all channels
```
Comment (Andrey1994) [inline OpenBCI_GUI/W_Focus.pde:226]: I think with channel selection UI this loop is not needed

File: javaEEGMetrics.pde [hunk 6]
```diff
@@ -0,0 +1,93 @@
+package brainflow.examples;
```
Comment (Andrey1994) [inline javaEEGMetrics.pde:1]: we can remove this file

**Review by Andrey1994 — APPROVED** [review]
looks good so far!

File: OpenBCI_GUI/W_Focus.pde (hunk 5 above)
Comment (retiutut) [reply OpenBCI_GUI/W_Focus.pde:226]: Right... should be able to just use toArray() good catch.

**Comment by retiutut** [comment]
@Andrey1994 Fixing the scrolling accuracy of the new Focus widget time-series graph will also unlock a fix for the same issue in the Spectrogram widget: https://github.com/OpenBCI/OpenBCI_GUI/issues/661

**Pull request merged**